### ISSUE TITLE
make each peer register as a producer in portforward

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3350]
|<img alt="rgl's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| rgl reported|
|-|-|
|Trac ID|trac#3350|
|Type|defect|
|Created|2008-07-12 12:19:55Z|
|Branch|https://github.com/twisted/twisted/tree/portforward-producers-3350|

The portforward protocol should register both peers as a producer of each one; this stops connections from filling the proxy memory when one side produces data at a higher rate than the other can consume.

Attachments:

* [portforward.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/741/741002b545bb31295fd1ae1d9bcdfcdd60276e9e/323701836e19b6a00fb5e9b947834e097ed6e7f5.patch) (791 bytes) - added by rgl on 2008-07-12 12:20:31Z - patch
* [portforward+test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/741/741002b545bb31295fd1ae1d9bcdfcdd60276e9e/bbe294d3860ab4cd22fb24142e8d197415b04ff0.patch) (2131 bytes) - added by verterok on 2011-08-31 13:26:53Z - portforward patch + unittest
* [portforward+better-test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/741/741002b545bb31295fd1ae1d9bcdfcdd60276e9e/46848763812347c98cca513935ff3176e0e5c4a9.patch) (3322 bytes) - added by verterok on 2011-09-13 19:44:18Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3350 3350
type__defect defect
reporter__rgl rgl
priority__high high
milestone__ 
branch__branches_portforward_producers_3350 branches/portforward-producers-3350
branch_author__rgl__verterok rgl, verterok
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1215865195000000 1215865195000000
changetime__1318786478000000 1318786478000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This looks like a great enhancement to make to the port forwarder code.  It needs unit tests, though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="verterok's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|verterok commented|
|-|-|

Hi,

I found this issue while deploying a new proxy, so here is a patch with a basic unittest for the change.

I'ld like to know if a integration test is needed, with integration I mean connecting to a "real" server and testing that the producers get paused/resumed as expected...It sounds like testing the producer/consumer stuff so I'm asking before investing on a potentially unneeded test :)

Cheers,

**Comment by twisted-trac** [comment]
|<img alt="verterok's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|verterok set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to verterok|
|-|-|

Thanks a lot for working on this verterok.  The patch looks pretty good, but I have a few comments about the unit test:

  1. Instead of clobbering `ProxyServer.connectionMade`, how about getting it to use a `twisted.test.proto_helpers.MemoryReactor` instance set up by the test, instead of the global reactor?  That way the test will exercise the actual `connectionMade` implementation still still avoid setting up a real TCP connection.
  1. The assertions could be augmented with a check of the `streaming` attribute on the `StringTransport` instances.  It needs to be `True` on both of them, since both transports are streaming producers.
  1. The test method should be `test_registerProducers` and the docstring can leave off the "Test that" prefix and just say the rest.

Thanks again!

**Comment by twisted-trac** [comment]
|<img alt="verterok's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|verterok removed owner|
|-|-|

Hi,

Attached a new version of the patch, (2) and (3) should be solved.
tried to use the twisted.test.proto_helpers.MemoryReactor but it doesn't call makeConnection, so I'm just using it to assert that portforward.ProxyServer.connectionMade calls reactor.connectTCP properly.

Regards,

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [32547]) Branching to 'portforward-producers-3350'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [32548]) Apply portforward+better-test.patch 

refs [#3350](https://github.com/twisted/twisted/issues/3350)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks!  Just a few more simple comments about the test:

  1. The test method name has a typo :)
  1. Test method docstrings should be phrased as "Foo does bar." instead of "Test foo does bar." or "Test that foo does bar."
  1. `proto_helpers.IPv4Address` is sneaky - `IPv4Address` actually comes from `twisted.internet.address` - using it on `proto_helpers` only works because `proto_helpers` imported it for its own use.  It would only be safe to use if `proto_helpers` put `IPv4Address` in its `__all__`, which it doesn't.
  1. This is pretty minor, but I find the test much easier to read if it uses symmetric naming for the client and server variables - that is, if "proxy" and "proxyTransport" are replaced with "server" and "serverTransport", to mirror the existing "client" and "clientTransport".
  1. There's no need to create the client protocol and call its `setPeer` method manually - the test goes on to grab the factory for this client connection out of the memory reactor.  The test can call `buildProtocol` on that to get the client protocol object.  This makes the ordering a little more realistic and shortens the test a little.

Apart from this, the change also needs a news fragment - this is a nice feature addition, so it should get mentioned in the news file when we release. :)

These are pretty minor points, so I'll make the changes and then merge the branch.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [32549]) Fix minor test issues; re-order the last couple bits, so that the client connection gets established using the real implementation instead of test fakes.

refs [#3350](https://github.com/twisted/twisted/issues/3350)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [32551]) Merge portforward-producers-3350

Author: rgl, verterok
Reviewer: exarkun
Fixes: [#3350](https://github.com/twisted/twisted/issues/3350)

Change `twisted.protocols.portforward` to use the producer/consumer APIs to
do flow control between the client and server parts of a forwarded connection.
This prevents the buffering of an unbounded amount of data from a faster
connection while waiting on a slower connection.
